### PR TITLE
Improvement: Demand v1alpha2 generic resource support list 

### PR DESCRIPTION
**Pull request by etherandrius**
This PR: 
- Upgrades CRD definition of demands from v1beta1 to v1.
- Adds new GPU field to v1alpha1 (This is required for conversion reasons)
image

- Extracts version definition for v1alpha1 (This is needed because v1alpha2 CRD has to include v1alpha1 version definition, for the time period where both versions are supported)
- Adds v1alpha2. The only type change is in demandUnit struct.
- Adds conversion code between the versions which follows this https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/conversion/conversion.go. 

For more context this tutorial on apiVersions is useful https://book.kubebuilder.io/multiversion-tutorial/tutorial.html

### CODE REVIEW

**Comment by etherandrius** [comment]
@rkaram @splucs for review

File: pkg/apis/scaler/v1alpha1/crd_demand.go [hunk 1]
```diff
@@ -49,109 +49,118 @@ var (
...
+								Type: "array",
+								Items: &v1.JSONSchemaPropsOrArray{
+									Schema: &v1.JSONSchemaProps{
+										Type:     "object",
+										Required: []string{"count", "cpu", "memory"},
+										Properties: map[string]v1.JSONSchemaProps{
+											"count":  {Type: "integer", Minimum: &oneFloat},
+											"cpu":    {Type: "string", MinLength: &oneInt},
+											"memory": {Type: "string", MinLength: &oneInt},
+											"gpu":    {Type: "string", MinLength: &oneInt},
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha1/crd_demand.go:97]: We won't support GPUs in v1alpha1 no?

File: pkg/apis/scaler/v1alpha1/types_demand.go [hunk 2]
```diff
@@ -58,6 +58,7 @@ type DemandStatus struct {
 type DemandUnit struct {
 	CPU    resource.Quantity `json:"cpu"`
 	Memory resource.Quantity `json:"memory"`
+	GPU    resource.Quantity `json:"gpu,omitempty"`
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha1/types_demand.go:61]: This API version is getting deprecated, I don't think we should be adding new things here.

File: pkg/apis/scaler/v1alpha1/crd_demand.go [hunk 3]
```diff
@@ -15,7 +15,7 @@
 package v1alpha1
 
 import (
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha1/crd_demand.go:18]: Doing this bump should have been done in another PR.

File: pkg/apis/scaler/v1alpha2/crd_demand.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	// AllDemandPhases is a list of all phases that a demand object could be in
+	AllDemandPhases = []string{
+		DemandPhaseEmpty,
+		DemandPhasePending,
+		DemandPhaseFulfilled,
+		DemandPhaseCannotFulfill,
+	}
+
+	// AllSupportedResources is a list of all resources that the demand object supports
+	AllSupportedResources = []corev1.ResourceName{
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/crd_demand.go:55]: Why do we need this? We should be able to support generic resources.

File: pkg/apis/scaler/v1alpha1/crd_demand.go [hunk 5]
```diff
@@ -49,109 +49,118 @@ var (
 	demandGroupResource        = demandGroupVersionResource.GroupResource()
 	oneFloat                   = float64(1)
 	oneInt                     = int64(1)
-	demandDefinition           = v1beta1.CustomResourceDefinition{
+	versionDefinition          = v1.CustomResourceDefinitionVersion{
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha1/crd_demand.go:52]: `v1alpha1VersionDefinition `

File: pkg/apis/scaler/v1alpha2/crd_demand.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// DemandPhaseEmpty is the state of a demand object when it is first created
+	DemandPhaseEmpty string = ""
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/crd_demand.go:25]: Can you make the `v1alpha1` consts point to these?

File: pkg/apis/scaler/v1alpha2/crd_demand.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+package v1alpha2
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/crd_demand.go:23]: Can you move these to the `types_demand.go` file?

File: pkg/apis/scaler/v1alpha2/types_demand.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+// Memory Returns the Memory demand if specified
+func (r *ResourceList) Memory() *resource.Quantity {
+	if val, ok := (*r)[ResourceMemory]; ok {
+		return &val
+	}
+	return &resource.Quantity{Format: resource.DecimalSI}
+}
+
+// NvidiaGPU Returns the GPU demand if specified
+func (r *ResourceList) NvidiaGPU() *resource.Quantity {
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/types_demand.go:95]: I don't think we should be adding anything specific to GPUs here.

**Review by splucs — CHANGES_REQUESTED** [review]

File: pkg/apis/scaler/v1alpha1/crd_demand.go (hunk 1 above)
Comment (etherandrius) [reply pkg/apis/scaler/v1alpha1/crd_demand.go:97]: see description, we need to add it for the conversion code.

File: pkg/apis/scaler/v1alpha1/types_demand.go (hunk 2 above)
Comment (etherandrius) [reply pkg/apis/scaler/v1alpha1/types_demand.go:61]: see above

File: pkg/apis/scaler/v1alpha2/crd_demand.go (hunk 6 above)
Comment (etherandrius) [reply pkg/apis/scaler/v1alpha2/crd_demand.go:25]: I can make the import v1alpha2 inside v1alpaha1 not the other way round. Will do it

File: pkg/apis/scaler/v1alpha2/types_demand.go (hunk 8 above)
Comment (etherandrius) [reply pkg/apis/scaler/v1alpha2/types_demand.go:95]: We should declare the resources that spark-scheduler supports. If this + resourceName was not declared users wouldn't know nvidia gpus are supported

File: pkg/apis/scaler/v1alpha2/crd_demand.go (hunk 4 above)
Comment (etherandrius) [reply pkg/apis/scaler/v1alpha2/crd_demand.go:55]: We should define what resources are supported first class by spark scheduler. 

With out this users won't know that `nvidia.com/gpu` is supported by spark scheduler
With out this users won't know that `glob-resource` is not supported spark scheduler

File: pkg/apis/scaler/v1alpha2/crd_demand.go (hunk 6 above)
Comment (splucs) [reply pkg/apis/scaler/v1alpha2/crd_demand.go:25]: I meant make v1alpha1 import v1alpha2.

File: pkg/apis/scaler/v1alpha1/conversion_demand.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/apis/scaler/v1alpha2"
+	werror "github.com/palantir/witchcraft-go-error"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+// ConvertTo converts from v1alpha1 to the storage version v1alpha2
+func (d *Demand) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*v1alpha2.Demand)
+
+	// ObjectMeta
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha1/conversion_demand.go:27]: These comments are unnecessary.

File: pkg/apis/scaler/v1alpha2/crd_demand.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+											"resources": {
+												Type: "object",
+												Properties: map[string]v1.JSONSchemaProps{
+													"cpu":            {Type: "string", MinLength: &oneInt},
+													"memory":         {Type: "string", MinLength: &oneInt},
+													"nvidia.com/gpu": {Type: "string", MinLength: &oneInt},
+												},
+											},
+											"count":  {Type: "integer", Minimum: &oneFloat},
+											"memory": {Type: "string", MinLength: &oneInt},
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/crd_demand.go:113]: Why do we need to explitly add memory?

File: pkg/apis/scaler/v1alpha2/crd_demand.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+	// DemandPhasePending is the state a demand object is in when Scaler has acknowledged it but has not yet taken
+	// action to fulfill the demand
+	DemandPhasePending string = "pending"
+	// DemandPhaseFulfilled is the state a demand object is in when Scaler has taken action and the action has completed
+	// to fulfill the demand. At this point, it is expected that there is capacity to meet the demand the object represents
+	DemandPhaseFulfilled string = "fulfilled"
+	// DemandPhaseCannotFulfill is the state a demand object is in when Scaler is unable to satisfy the demand. This is
+	// possible if the demand contains a single unit that is larger than the instance group is configured to use, or if
+	// the instance group has reached its maximum capacity and cannot allocate more
+	DemandPhaseCannotFulfill string = "cannot-fulfill"
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/crd_demand.go:35]: Can you make these a proper "DemandPhase" type?

**Review by splucs — CHANGES_REQUESTED** [review]

File: pkg/apis/scaler/v1alpha2/crd_demand.go (hunk 10 above)
Comment (etherandrius) [reply pkg/apis/scaler/v1alpha2/crd_demand.go:113]: that's a mistake. fixing.

File: pkg/apis/scaler/v1alpha2/types_demand.go [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+// DemandList is a list of Demand resources
+type DemandList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []Demand `json:"items"`
+}
+
+// CPU Returns the CPU demand if specified
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/types_demand.go:87]: Lower case return, end comment with period.

File: pkg/apis/scaler/v1alpha2/types_demand.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+func (r *ResourceList) CPU() *resource.Quantity {
+	if val, ok := (*r)[ResourceCPU]; ok {
+		return &val
+	}
+	return &resource.Quantity{Format: resource.DecimalSI}
+}
+
+// Memory Returns the Memory demand if specified
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/types_demand.go:95]: Lower case return, end comment with period.Lower case return, end comment with period.

File: pkg/apis/scaler/v1alpha2/types_demand.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+func (r *ResourceList) Memory() *resource.Quantity {
+	if val, ok := (*r)[ResourceMemory]; ok {
+		return &val
+	}
+	return &resource.Quantity{Format: resource.DecimalSI}
+}
+
+// NvidiaGPU Returns the GPU demand if specified
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/types_demand.go:103]: Lower case return, end comment with period.

**Review by splucs — CHANGES_REQUESTED** [review]

File: pkg/apis/scaler/v1alpha1/conversion_demand.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/apis/scaler/v1alpha2"
+	werror "github.com/palantir/witchcraft-go-error"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+// ConvertTo converts from v1alpha1 to the storage version v1alpha2
+func (d *Demand) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*v1alpha2.Demand)
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha1/conversion_demand.go:25]: You shouldn't panic if this is not the case. Return an error instead.

File: pkg/apis/scaler/v1alpha1/conversion_demand.go [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+		})
+	}
+	dst.Spec.Units = dstUnits
+
+	return nil
+}
+
+// ConvertFrom converts from storage version v1alpha2 to v1alpha1
+func (d *Demand) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*v1alpha2.Demand)
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha1/conversion_demand.go:53]: You shouldn't panic if this is not the case. Return an error instead.

File: pkg/apis/scaler/v1alpha2/crd_demand.go [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+								Type: "array",
+								Items: &v1.JSONSchemaPropsOrArray{
+									Schema: &v1.JSONSchemaProps{
+										Type:     "object",
+										Required: []string{"count", "resources"},
+										Properties: map[string]v1.JSONSchemaProps{
+											"resources": {
+												Type: "object",
+												Properties: map[string]v1.JSONSchemaProps{
+													"cpu":            {Type: "string", MinLength: &oneInt},
```
Comment (splucs) [inline pkg/apis/scaler/v1alpha2/crd_demand.go:92]: Can we use the consts here?

**Review by splucs — CHANGES_REQUESTED** [review]

File: pkg/apis/scaler/v1alpha1/conversion_demand.go (hunk 15 above)
Comment (etherandrius) [reply pkg/apis/scaler/v1alpha1/conversion_demand.go]: From what I looked at the conversion code this should be a safe conversion. However, will still add the errors just in case

**Review by splucs — APPROVED** [review]
